src/assembly.h

```diff
@@ -383,32 +383,15 @@ static __inline Word64 MADD64(Word64 sum64, int x, int y)
         return u.w64;
 }
 
-/* Ken's trick: clips to [-32768, 32767] */
-//sign = x >> 31;
-//if (sign != (x >> 15))
-//    x = sign ^ ((1 << 15) - 1);
-__attribute__((__always_inline__)) static __inline short SAR64_Clip(Word64 x, int n)
+// Shift the 64-bit value right by 26 bits and return the lower 16-bits
+__attribute__((__always_inline__)) static __inline short SAR64_Clip(Word64 x)
 {
   unsigned int xLo = (unsigned int) x;
   int xHi = (int) (x >> 32);
-  int nComp = 32-n;
-  int tmp;
-  // Shortcut: n is always < 32.
-  __asm__ __volatile__( "lsl %2, %0, %3\n\t"  // tmp <- xHi<<(32-n)
-                        "lsr %1, %1, %4\n\t"  // xLo <- xLo>>n
-                        "orr %1, %2\n\t"      // xLo <= xLo || tmp
-// Uncomment this part if you really need it to saturate the output to 16-bits
-// This didn't appear necessary because the temp data is shifted right by 26 bits
-// and doesn't grow large enough to overflow a 16-bit signed value
-//                        "asr %2, %1, #31\n\t" // get sign in tmp
-//                        "asr %0, %1, #15\n\t" // use xHi as tmp2
-//                        "mov %3, #-1\n\t" // prep constant 0x7fff
-//                        "lsr %3, #17\n\t"
-//                        "cmp %2, %0\n\t"      // if (sign != (x >> 15))
-//                        "it ne\n\t"
-//                        "eorne %1, %3\n\t"
-                        : "+&r" (xHi), "+r" (xLo), "=&r" (tmp)
-                        : "r" (nComp), "r" (n) );
+  __asm__ __volatile__(
+                        "lsr %1, %1, #26\n\t"  // xLo <- xLo>>n
+                        "orr %1, %1, %0, lsl #6\n\t"      // xLo <= xLo || (xHi << 6)
+                        : "+&r" (xHi), "+r" (xLo) );
```

Comment:
If written like this, the code will automatically adapt if n is constant and use the more efficient sequence:
```
__attribute__((__always_inline__)) static __inline short SAR64_Clip(Word64 x, int n) {
    unsigned int xLo = (unsigned int) x;
    int xHi = (int) (x >> 32);
    if(__builtin_constant_p(n)) {
        __asm__ __volatile__(
            "lsr %1, %1, %2\n\t"         // xLo <- xLo>>(32-n)
            "orr %1, %1, %0, lsl %3\n\t" // xLo <= xLo || (xHi << n)
            : "+&r" (xHi), "+r" (xLo)
            : "M" (n), "M" (32-n) );
    } else {
        int nComp = 32-n; 
        int tmp;
        __asm__ __volatile__(
            "lsl %2, %0, %3\n\t"         // tmp <- xHi<<(32-n)
            "lsr %1, %1, %4\n\t"         // xLo <- xLo>>n
            "orr %1, %2\n\t"             // xLo <= xLo || tmp
            : "+&r" (xHi), "+r" (xLo), "=&r" (tmp)
            : "r" (nComp), "r" (n) );
    }
    return( (short)xLo );
}
```
Otherwise, it might be good to document it (e.g., by putting it in the function name?) that this is a constant shift of 26 bits.

Is there a missing implementation (or two?) of this function for the other CPU and compiler choices?

Reply:
Good idea about the constant vs variable shift amount. I've written a lot of assembly language over the years, but avoided inline assembly because it would get mangled by the compiler into something different. External asm files were safe from compiler changes.

I wrote the x86 asm version of the function, but perhaps I missed something. If you're happy, then please do the merge.

Reply:
In the version that exists now, it looks like you have incompatible SAR64_clip between x86 and arm.
```
+static __inline short SAR64_Clip(Word64 x, int n)
+__attribute__((__always_inline__)) static __inline short SAR64_Clip(Word64 x)
```
I don't know how to build the x86 version, but this seems fishy to me.  Please let me know what's up.